test/logstream/kubelogs.go

```diff
@@ -49,54 +53,99 @@ var _ streamer = (*kubelogs)(nil)
+				return err
+			}
+			defer stream.Close()
+			// Read this container's stream.
+			scanner := bufio.NewScanner(stream)
+			for scanner.Scan() {
+				k.handleLine(scanner.Bytes())
+			}
+			// Pods get killed with chaos duck, so logs might end
+			// before the test does.
+		})
+	}
+}
+
+func (k *kubelogs) watchPods(t test.TLegacy, eg *errgroup.Group) {
+	wi, err := k.kc.Kube.CoreV1().Pods(k.namespace).Watch(metav1.ListOptions{})
+	if err != nil {
+		t.Logf("Logstream knative pod watch failed, logs might be missing: %v", err)
+		return
+	}
+	go func() {
+		for ev := range wi.ResultChan() {
+			p := ev.Object.(*corev1.Pod)
+			t.Logf("Pod event %v for %s", ev.Type, p.Name)
+			switch ev.Type {
+			case watch.Deleted:
+				k.watchedPods.Delete(p.Name)
+			case watch.Added, watch.Modified:
+				if k.watchedPods.Has(p.Name) {
+					t.Log("Already watching pod", p.Name)
+					continue
+				}
+				// If pod is ready then start the logs.
+				if p.Status.Phase == corev1.PodRunning && p.DeletionTimestamp == nil {
+					for _, cond := range p.Status.Conditions {
+						if cond.Type == corev1.PodReady && cond.Status == corev1.ConditionTrue {
+							t.Log("The new knative pod is ready", p.Name)
+							k.startForPod(eg, p)
+							break
+						}
+					}
+				}
+				t.Log("Not ready yet pod", p.Name)
+			}
+		}
+	}()
+}
+
 func (k *kubelogs) init(t test.TLegacy) {
 	k.keys = make(map[string]logger)
 
 	kc, err := test.NewKubeClient(test.Flags.Kubeconfig, test.Flags.Cluster)
 	if err != nil {
 		t.Error("Error loading client config", "error", err)
+		return
 	}
+	k.kc = kc
 
 	// List the pods in the given namespace.
 	pl, err := kc.Kube.CoreV1().Pods(k.namespace).List(metav1.ListOptions{})
 	if err != nil {
 		t.Error("Error listing pods", "error", err)
 	}
 
+	k.watchedPods = make(sets.String, len(pl.Items))
+
 	eg := errgroup.Group{}
 	for _, pod := range pl.Items {
-		// Grab data from all containers in the pods.  We need this in case
-		// an envoy sidecar is injected for mesh installs.  This should be
-		// equivalent to --all-containers.
-		for _, container := range pod.Spec.Containers {
-			// Required for capture below.
-			pod, container := pod, container
-			eg.Go(func() error {
-				options := &corev1.PodLogOptions{
-					Container: container.Name,
-					// Follow directs the api server to continuously stream logs back.
-					Follow: true,
-					// Only return new logs (this value is being used for "epsilon").
-					SinceSeconds: ptr.Int64(1),
-				}
-
-				req := kc.Kube.CoreV1().Pods(k.namespace).GetLogs(pod.Name, options)
-				stream, err := req.Stream()
-				if err != nil {
-					return err
-				}
-				defer stream.Close()
-				// Read this container's stream.
-				scanner := bufio.NewScanner(stream)
-				for scanner.Scan() {
-					k.handleLine(scanner.Bytes())
-				}
-				return fmt.Errorf("logstream completed prematurely for %s/%s: %w",
-					pod.Name, container.Name, scanner.Err())
-			})
-		}
+		pod := pod
+		k.startForPod(&eg, &pod)
```

Comment:
Do we still need this loop, or can it be subsumed by watchPods?

Reply:
yeah, I think so.